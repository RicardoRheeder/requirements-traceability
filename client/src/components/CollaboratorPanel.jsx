import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { setUserColorObject } from '../redux/stores/common/actions'

import { CollaboratorIcon } from './'

const testUserObjects = [
  { userID: '12312', username: 'ricardo' },
  { userID: '54564', username: 'Faiz' },
  { userID: '84512', username: 'Devin' },
]

const getRandColor = (brightness) => {
  // Six levels of brightness from 0 to 5, 0 being the darkest
  var rgb = [Math.random() * 256, Math.random() * 256, Math.random() * 256]
  var mix = [brightness * 51, brightness * 51, brightness * 51] //51 => 255/5
  var mixedrgb = [rgb[0] + mix[0], rgb[1] + mix[1], rgb[2] + mix[2]].map(
    function (x) {
      return Math.round(x / 2.0)
    }
  )
  return 'rgb(' + mixedrgb.join(',') + ')'
}

const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min) + min) //The maximum is exclusive and the minimum is inclusive
}

export default function CollaboratorPanel() {
  const dispatch = useDispatch()
  const userColorObject = useSelector((state) => state.common.userColorObject)

  const generateUserIcons = (userStruct) => {
    return userStruct.map(({ userID, username }, i) => {
      let randomColor

      if (userColorObject[username]) {
        randomColor = userColorObject[username]
      } else {
        randomColor = getRandColor(getRandomInt(3, 6))
        dispatch(setUserColorObject({[username]: randomColor}))
      }

      return (
        <CollaboratorIcon key={i} username={username} color={randomColor} />
      )
    })
  }

  return (
    <span className="collaborator-panel">
      {generateUserIcons(testUserObjects)}
    </span>
  )
}
