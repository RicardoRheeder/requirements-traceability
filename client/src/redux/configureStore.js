import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

// restricting extension in production using process.env.NODE_ENV which is set 'production'
import { composeWithDevTools } from 'redux-devtools-extension/logOnlyInProduction';

import store from './stores';

const enhancer = (process.env.NODE_ENV == 'development')
               ? composeWithDevTools(applyMiddleware(thunk))
               : applyMiddleware(thunk);

export default () => createStore(store, enhancer);
