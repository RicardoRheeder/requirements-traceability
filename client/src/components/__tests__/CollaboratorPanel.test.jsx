import React from "react"
import Enzyme, { shallow, mount} from "enzyme"
import CollaboratorPanel from "../CollaboratorPanel"
import Adapter from "enzyme-adapter-react-16"
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"


const mockStore = configureStore([])
Enzyme.configure({adapter: new Adapter() });

describe("CollaboratorPanel", () => {
    let store;
    beforeEach(() => {
        store = mockStore({
            document: {
                documents: [{title: 'testdoc', versions: [], tree: null}],
                versions: [],
                current_doc: {collaborators: [
                    { _id: '111', username: 'testUser1' },
                    { _id: '222', username: 'testUser2' },
                  ]}
            },
            common: {
                selectedDocumentPanelObject: {title: 'testdoc'}
            },
            user: {
                notifications: []
            }
            
        });
    });
    test("renders", () => {
        const wrapper = shallow(<Provider store = {store}><CollaboratorPanel/></Provider>);
        expect(wrapper.exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });
    test("renders with children", () => {
        const wrapper = mount(<Provider store = {store}><CollaboratorPanel/></Provider>);
        expect(wrapper.exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });
});