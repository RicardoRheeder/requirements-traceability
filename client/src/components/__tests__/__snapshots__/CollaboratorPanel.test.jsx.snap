// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollaboratorPanel renders 1`] = `ShallowWrapper {}`;

exports[`CollaboratorPanel renders with children 1`] = `ReactWrapper {}`;
